From dee2ef4f33a940ebb39a8fcaedbb20c4201f7892 Mon Sep 17 00:00:00 2001
From: Quanyang Wang <quanyang.wang@windriver.com>
Date: Mon, 26 Jun 2023 08:48:28 +0800
Subject: [PATCH] pfe-slave: add mutex_unlock in pfe_emac_destroy

Calling mutex_destroy without mutex_unlock "emac->mutex" will trigger the warning:

	[  884.807088] =========================
	[  884.807091] WARNING: held lock freed!
	[  884.807095] 6.1.34-rt10-yocto-preempt-rt #1 Tainted: G        W  O
	[  884.807100] -------------------------
	[  884.807102] modprobe/2039 is freeing memory ffffff8810ed1600-ffffff8810ed16ff, with a lock still held there!
	[  884.807110] ffffff8810ed1670 (mutex#23){+.+.}-{3:3}, at: pfe_emac_destroy+0x2c/0x50 [pfeng_slave]
	[  884.807195] 2 locks held by modprobe/2039:
	[  884.807200]  #0: ffffff88030ea0f0 (&dev->mutex){....}-{3:3}, at: device_release_driver_internal+0x48/0x270
	[  884.807229]  #1: ffffff8810ed1670 (mutex#23){+.+.}-{3:3}, at: pfe_emac_destroy+0x2c/0x50 [pfeng_slave]
	[  884.807289]
	[  884.807289] stack backtrace:
	[  884.807295] CPU: 4 PID: 2039 Comm: modprobe Tainted: G        W  O       6.1.34-rt10-yocto-preempt-rt #1
	[  884.807302] Hardware name: Freescale S32G399A (DT)
	[  884.807307] Call trace:
	[  884.807309]  dump_backtrace.part.0+0xcc/0xe0
	[  884.807321]  show_stack+0x20/0x30
	[  884.807327]  dump_stack_lvl+0xac/0xf0
	[  884.807338]  dump_stack+0x18/0x34
	[  884.807346]  debug_check_no_locks_freed+0x128/0x15c
	[  884.807357]  slab_free_freelist_hook.constprop.0+0xa0/0x1a0
	[  884.807368]  __kmem_cache_free+0x78/0x1f0
	[  884.807375]  kfree+0xe0/0x1f0
	[  884.807385]  oal_mm_free+0x18/0x24 [pfeng_slave]
	[  884.807434]  pfe_emac_destroy+0x3c/0x50 [pfeng_slave]
	[  884.807482]  pfe_platform_destroy_emac+0x54/0x80 [pfeng_slave]
	[  884.807530]  pfe_platform_remove+0x2c/0xb4 [pfeng_slave]
	[  884.807579]  pfeng_drv_remove+0x78/0x210 [pfeng_slave]
	[  884.807627]  platform_remove+0x5c/0x74
	[  884.807635]  device_remove+0x54/0x8c
	[  884.807643]  device_release_driver_internal+0x200/0x270
	[  884.807651]  driver_detach+0x58/0xa0
	...

Add mutex_unlock to fix this warning.

Upstream-Status: Pending

Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
---
 sw/pfe_platform/src/pfe_emac_slave.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/sw/pfe_platform/src/pfe_emac_slave.c b/sw/pfe_platform/src/pfe_emac_slave.c
index ad8358f..e9a18c3 100644
--- a/sw/pfe_platform/src/pfe_emac_slave.c
+++ b/sw/pfe_platform/src/pfe_emac_slave.c
@@ -152,6 +152,11 @@ void pfe_emac_destroy(pfe_emac_t *emac)
 			NXP_LOG_DEBUG("Mutex lock failed\n");
 		}
 
+		if (EOK != oal_mutex_unlock(&emac->mutex))
+		{
+			NXP_LOG_DEBUG("Mutex unlock failed\n");
+		}
+
 		/*	Destroy mutex */
 		(void)oal_mutex_destroy(&emac->mutex);
 
-- 
2.35.5

